kadena-io/chainweb-node
# Simple Remote Miner Communication

This PR introduces the `--remote-miner` flag. If specified at least once, `chainweb-node` will forego running its own mining "inner loop", and instead make remote calls to hosts provided by the aforementioned flag.

Since the existing `chainweb-miner` only provides the tight inner loop and knows nothing else about Chainweb, we follow the "push-work-pull-results" model of fetching mined blocks from the remote processes. This may or may not be optimal, but it's a start.

Note also for @larskuhtz that I haven't yet attempted a unifying typeclass for "inner loops". Right now the inner loop implementations live in `Chainweb.Miner.Miners`.

### Testing this PR

1. Install `chainweb-node` and `chainweb-miner` by your usual means (`stack install`, etc.).
2. Run `chainweb-miner`as-is. You should see:
```
colin@yumi ~/c/h/chainweb> chainweb-miner
ClientArgs 
    { cores = 1
    , version = development
    , port = 8081
    } 
```
3. Ping the miner:
```
colin@yumi ~> curl localhost:8081/env
{"version":"development","cores":1,"port":8081}
```
4. Run `chainweb-node --chainweb-version=development --remote-miner=localhost:8081` and watch it work.
5. Stop `chainweb-node` and run a second miner: `chainweb-miner --port 8082`
6. Run `chainweb-node --chainweb-version=development --remote-miner=localhost:8081 --remote-miner=localhost:8082` to communicate with both clients at once!
